File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        hmap = {}
        
        hmap[0] = 1
        
        total = 0
        prefixSum = 0
        
        for i in range(len(nums)):
            prefixSum += nums[i]
            
            if prefixSum - k in hmap:
                total += hmap[prefixSum-k]
            if prefixSum in hmap:
                hmap[prefixSum] += 1
            else:
                hmap[prefixSum] = 1
        
        return total
        

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int tCount = 0;
        int pSum = 0;
        unordered_map<int, int> hmap;
        hmap[0] = 1;
        
        for( int i = 0; i < nums.size(); i++) {
            pSum += nums[i];
            
            // how many prefixes sum to prefixsum - k
            if( hmap.find(pSum - k) != hmap.end() )
                tCount += hmap[pSum - k];
            // Update hmap
            if ( hmap.find( pSum ) != hmap.end() )
                hmap[pSum]++;
            else
                hmap[pSum] = 1;
        }
        
        return tCount;
    }
};